code/go/internal/validator/semantic/validate_kibana_no_dangling_object_ids.go

```diff
@@ -4,14 +4,170 @@
+			errs = append(errs, fmt.Errorf("unable to create reference from file [%s]: %w", fsys.Path(filePath), err))
+		}
+
+		installedIDs = append(installedIDs, currentReference)
+
+		referencedObjects, err := getReferencesListFromCurrentObject(objectFile, fsys.Path(filePath))
+		if err != nil {
+			errs = append(errs, fmt.Errorf("unable to create referenced objects from file [%s]: %w", fsys.Path(filePath), err))
+			continue
+		}
+
+		referencedIDs = append(referencedIDs, referencedObjects...)
+	}
+
+	if len(referencedIDs) == 0 {
+		return errs
+	}
+
+	for _, reference := range referencedIDs {
+		// look for installed IDs
+		found := false
+		for _, installed := range installedIDs {
+			if reference.objectID != installed.objectID {
+				continue
+			}
+			if reference.objectType != installed.objectType {
+				continue
+			}
+
+			found = true
+		}
+		if !found {
+			errs = append(errs, fmt.Errorf("file \"%s\" is invalid: dangling reference found: %s (%s)", reference.filePath, reference.objectID, reference.objectType))
+		}
+	}
+
+	return errs
+}
+
+func getCurrentObjectReference(asset pkgpath.File, filePath string) (objectReference, error) {
+	var reference objectReference
+
+	valueID, err := asset.Values("$.id")
+	if err != nil {
+		return reference, fmt.Errorf("unable to get ID field : %w", err)
+	}
+	stringValueID, ok := valueID.(string)
+	if !ok {
+		return reference, fmt.Errorf("expect value ID to be a string: %w", err)
+	}
+
+	valueType, err := asset.Values("$.type")
+	if err != nil {
+		return reference, fmt.Errorf("unable to get Type field : %w", err)
+	}
+	stringValueType, ok := valueType.(string)
+	if !ok {
+		return reference, fmt.Errorf("expect value Type to be a string: %w", err)
+	}
+
+	reference = objectReference{
+		objectID:   stringValueID,
+		objectType: stringValueType,
+		filePath:   filePath,
+	}
+
+	return reference, nil
+}
+
+func getReferencesListFromCurrentObject(asset pkgpath.File, filePath string) ([]objectReference, error) {
+	referencedIDs := []objectReference{}
+	objectReferences, err := asset.Values(`$.references`)
+	if err != nil {
+		// no references key in dashboard json
+		return referencedIDs, nil
+	}
+
+	references, err := filterReferences(objectReferences, exceptionAssets)
+	if err != nil {
+		return nil, fmt.Errorf("error getting references: %w", err)
+	}
+
+	if len(references) == 0 {
+		return referencedIDs, nil
+	}
+
+	for _, reference := range references {
+		referencedIDs = append(referencedIDs, objectReference{
+			objectID:   reference.ID,
+			objectType: reference.Type,
+			filePath:   filePath,
+		})
+	}
+
+	return referencedIDs, nil
+}
+
+func filterReferences(val interface{}, exceptions []string) ([]reference, error) {
+	allReferences, err := toReferenceSlice(val)
+	if err != nil {
+		return []reference{}, fmt.Errorf("unable to convert references: %w", err)
+	}
+
+	if len(allReferences) == 0 {
+		return []reference{}, nil
+	}
+
+	var references []reference
+	for _, reference := range allReferences {
```

Comment:
Nit. Also here maybe you can use some `slices` helper.

Reply:
Changed to use `slices.Contains` :+1: